adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+------------------------ MODULE query_deduplication ------------------------
+
+
+EXTENDS Integers, Sequences, FiniteSets, TLC
+
+CONSTANT Proc
+CONSTANT Keys
+(*
+
+\* from https://probablydance.com/2020/10/31/using-tla-in-the-real-world-to-understand-a-glibc-bug/
+--algorithm Spinlock 
+{
+  variables
+    cache_lock = 0;
+    wait_map_lock = 1;
+    locks = [i \in {cache_lock, wait_map_lock} |-> FALSE];
+    \* set to false for every possible key at first
+    cache = [ i \in Keys |-> FALSE ];
+    wait_map = [ i \in Keys |-> {} ];
+    channels = [i \in Proc |-> FALSE ];
+    tested_keys = {};
+ 
+   define {
+
+    MutualExclusion == \A i, j \in Proc : 
+                         (i # j) => ~ /\ pc[i] = "cs"
+                                      /\ pc[j] = "cs"
+                                      
+    DeadlockFreedom == 
+        \A i \in Proc : 
+          (pc[i] = "exchange") ~> (\E j \in Proc : pc[j] = "cs")
+    
+    
+    \*AllFinished == <>(\A proc \in Proc: pc[proc] = "Done")
+    \*AllKeysLoaded == <>(\A k \in tested_keys: cache[k])
+  };
+
+  procedure lock(idx = 0)
```

Comment:
this is the lock procedure. We call `await` on the condition that `locks[idx]` be false. `exchange:` and `check:` are "labels". Each label is a state at which a process can be stopped and restarted later. This is how TLA tests different orderings, it tries combinations of state transitions.

All of the expressions under a label are executed atomically. So in `check`, we know that the value of `locks[idx]` will not change between `if(~locks[idx])` and `locks[idx] := TRUE;`.
But it can change between `await ~locks[idx]`, which executes under the `exchange` state, and `if(~locks[idx])` which executes under `check`.
This is how the mutex is implemented: when it is released (`locks[idx] := FALSE`), all of the process stuck on `await` can be woken up, but the first one will take the lock, all the other one will see it set to `TRUE` and execute `goto exchange`